src/frontend/src/optimizer/plan_node/logical_agg.rs

```diff
@@ -701,21 +701,67 @@ impl ToBatch for LogicalAgg {
 
 impl ToStream for LogicalAgg {
     fn to_stream(&self) -> Result<PlanRef> {
+        let input = self.input();
+        // simple-agg
         if self.group_keys().is_empty() {
-            Ok(StreamSimpleAgg::new(
-                self.clone_with_input(
-                    self.input()
-                        .to_stream_with_dist_required(&RequiredDist::single())?,
-                ),
-            )
-            .into())
+            let mut has_min_or_max_agg_calls = false;
+            for c in &self.agg_calls {
+                match c.agg_kind {
+                    AggKind::Min | AggKind::Max => {
+                        has_min_or_max_agg_calls = true;
+                        break;
+                    }
+                    _ => continue,
+                }
+            }
+
+            let input_stream = input.to_stream()?;
+            let is_append_only = input_stream.append_only();
+            let input_distribution = input_stream.distribution();
+
+            // simple 2-phase-agg
+            if input_distribution.satisfies(&RequiredDist::AnyShard)
+                // TODO: non-append-only min/max is not supported yet,
+                // since min/max is a table rather than single value for local agg.
+                // see: https://github.com/singularity-data/risingwave/issues/2997#issuecomment-1156139576
+                && (is_append_only || !has_min_or_max_agg_calls)
+            {
+                // partial agg
+                let partial_agg_plan =
+                    StreamLocalSimpleAgg::new(self.clone_with_input(input_stream)).into();
+
+                // insert exchange
+                let exchange = StreamExchange::new(partial_agg_plan, Distribution::Single).into();
```

Comment:
You don't need to insert exchange manually. As long as you set global agg to `Single` distribution, there will be a phase to insert exchange afterwards IIRC.

Reply:
In some places `enforce_if_not_satisfies` will be called on stream plan nodes, you can have a try not manually inserting the exchange.

Reply:
Okay I call `enforce_if_not_satisfies` for more declarative approach.